man/dmrecv.1
.TH "DMRECV" "1" "" "Version 2.0.0" "User Commands"
.SH NAME
dmrecv \(en receives observations and logs from POSIX message queue
.SH SYNOPSIS
\f[B]dmrecv\f[R] \-\-help
.PP
\f[B]dmrecv\f[R] \-\-version
.PP
\f[B]dmrecv\f[R] \-\-\f[B]type\f[R] [log|observ] \-\-\f[B]format\f[R]
[block|csv|jsonl|nml] [\-\-\f[B]name\f[R] \f[I]name\f[R]]
[\-\-\f[B]output\f[R] \f[I]file\f[R]] [\-\-\f[B]response\f[R]
\f[I]name\f[R]] [\-\-\f[B]node\f[R] \f[I]id\f[R]] [\-\-\f[B]logger\f[R]
\f[I]name\f[R]] [\-\-\f[B]forward\f[R]] [\-\-\f[B]replace\f[R]]
[\-\-\f[B]verbose\f[R]]
.PP
\f[B]dmrecv\f[R] \-\-\f[B]config\f[R] \f[I]file\f[R] [\-\-\f[B]name\f[R]
\f[I]name\f[R]]
.SH DESCRIPTION
The \f[B]dmrecv\f[R] program listens to the POSIX message queue of its
name and writes received observations and logs to \f[I]stdout\f[R],
file, or named pipe; in ASCII block, CSV, JSON, or Fortran 95 Namelist
format.
By default, the serialised data is appended to the end of the output
file.
If argument \f[CR]\-\-replace\f[R] is passed, the file will be replaced
consecutively.
.PP
Received observations are not forwarded to the next specified receiver,
unless argument \f[CR]\-\-forward\f[R] is set.
If no receivers are defined or left, the observation will be discarded
after output.
.PP
The output format \f[CR]block\f[R] is available for observation data
only, and requires a valid response name.
Received observations will be searched for this response and converted
to data point type if found.
The data point is printed in ASCII block format.
.PP
If output format JSON Lines is selected, observations and logs are
written sequentially as JSON objects to file or \f[I]stdout\f[R],
separated by new line.
Use \f[I]jq(1)\f[R] to convert a file \f[CR]input.jsonl\f[R] of JSON
objects to a JSON array in \f[CR]output.json\f[R]:
.IP
.EX
$ jq \-s \(aq.\(aq input.jsonl > output.json
.EE
.PP
The program settings are passed through command\-line arguments or an
optional configuration file.
The arguments overwrite settings from file.
.SH OPTIONS
.TP
\f[B]\-\-config\f[R], \f[B]\-c\f[R] \f[I]file\f[R]
Path to configuration file.
.TP
\f[B]\-\-debug\f[R], \f[B]\-D\f[R]
Forward logs messages of level \f[CR]LL_DEBUG\f[R] via IPC (if logger is
set).
.TP
\f[B]\-\-format\f[R], \f[B]\-f\f[R] [block|csv|jsonl|nml]
Output format, either ASCII block, CSV, JSON Lines, or Namelist.
Format \f[CR]block\f[R] is allowed for type \f[CR]observ\f[R] only.
.TP
\f[B]\-\-forward\f[R], \f[B]\-F\f[R]
Forward each observation to its next specified receiver.
.TP
\f[B]\-\-help\f[R], \f[B]\-h\f[R]
Print available command\-line arguments and quit.
.TP
\f[B]\-\-logger\f[R], \f[B]\-l\f[R] \f[I]name\f[R]
Name of logger.
If set, sends logs to \f[I]dmlogger(1)\f[R] process of given name.
.TP
\f[B]\-\-name\f[R], \f[B]\-n\f[R] \f[I]name\f[R]
Name of configuration and POSIX message queue (default is
\f[CR]dmrecv\f[R]).
.TP
\f[B]\-\-node\f[R], \f[B]\-N\f[R] \f[I]id\f[R]
Optional node id.
.TP
\f[B]\-\-output\f[R], \f[B]\-o\f[R] \f[I]file\f[R]
Path of output file.
Prints serialised data to \f[I]stdout\f[R] if \f[CR]\-\f[R] or not set.
.TP
\f[B]\-\-replace\f[R], \f[B]\-r\f[R]
Replace output file instead of appending data.
.TP
\f[B]\-\-response\f[R], \f[B]\-R\f[R] \f[I]name\f[R]
Name of the observation response to output (required for format
\f[CR]block\f[R]).
.TP
\f[B]\-\-type\f[R], \f[B]\-t\f[R] [log|observ]
Data type to receive, either log or observation.
.TP
\f[B]\-\-verbose\f[R], \f[B]\-V\f[R]
Print log messages to \f[I]stderr\f[R].
.TP
\f[B]\-\-version\f[R], \f[B]\-v\f[R]
Print version information and quit.
.SH EXIT STATUS
.TP
\f[B]0\f[R]
Success.
Program terminated normally.
.TP
\f[B]1\f[R]
Failure.
Program execution failed.
.SH ENVIRONMENT
.TP
\f[B]DM_LOGGER\f[R]
Name of logger instance to send logs to.
.TP
\f[B]NO_COLOR\f[R]
Disable ANSI colour output.
.SH EXAMPLE
Write received log messages in CSV format to file
\f[CR]/tmp/logs.csv\f[R]:
.IP
.EX
$ dmrecv \-n dmrecv \-t log \-f csv \-o /tmp/logs.csv
.EE
.PP
Output observations in JSON format to \f[I]stdout\f[R]:
.IP
.EX
$ dmrecv \-n dmrecv \-t observ \-f json
.EE
.PP
Output observation responses of name \f[CR]x\f[R] in ASCII block format
to \f[I]stdout\f[R]:
.IP
.EX
$ dmrecv \-n dmrecv \-t observ \-f block \-R x
.EE
.SH SEE ALSO
\f[I]dmsend(1)\f[R]


CHANGELOG.md
# Change Log

All notable changes to the project will be documented in this file.

## [2.0.0] – Unreleased

## Library

* `Added` error codes
* `Added` _fortran-nng_ interface bindings to NNG to `vendor/`
* `Added` module `dm_arg_parser` (outsourced from `dm_arg`)
* `Added` module `dm_group` for observation groups
* `Added` module `dm_ipc` for socket access
* `Added` module `dm_ipc_async` for asynchronous I/O
* `Added` module `dm_ipc_disco` for service discovery
* `Added` module `dm_ipc_message` for message handling
* `Added` module `dm_ipc_mutex` for mutex access
* `Added` module `dm_ipc_thread` for thread access
* `Added` module `dm_ipc_type` of IPC enumerators
* `Added` module `dm_job_list` (outsourced from `dm_job`)
* `Changed` names of modules `dm_fifo`, `dm_mqueue`, `dm_mqueue_util`, `dm_pipe`,
  `dm_sem`, `dm_signal`, `dm_thread`, `dm_tty` to include `posix` prefix
* `Changed` name of module `dm_system` to `dm_posix`
* `Changed` database schema due to new data models
* `Changed` default database suffix from `.sqlite` to `.db`
* `Changed` GeoCOM API for Lua to accept prototype observation as first argument
* `Changed` model of observation data structure `dm_observ`
* `Deleted` model of request data strucutre `dm_request`

## Programs

* `Changed` response value units in `dmved`

## Documentation

* Converted user guide from AsciiDoc to Markdown
* Converted man pages from AsciiDoc to Markdown

## [1.0.0] – Unreleased

## Library

* `Added` Bessel, Butterworth, Chebyshev filters to `dm_filter`

## [0.9.9] – 2026-01-11

### Library

* `Added` module `dm_coord` for coordinate transformation
* `Added` module `dm_filter` for filtering
* `Added` module `dm_ghostscript` for Ghostscript access
* `Added` module `dm_netstring` for netstring parsing
* `Added` module `dm_statistics` of statistics functions
* `Added` file tree size function to module `dm_file`
* `Added` file modification date/time to function `dm_file_touch()` in `dm_file`
* `Added` format `FORMAT_TSV` for tab-separated values to `dm_format`
* `Added` swap routines to `dm_util`
* `Changed` error code order for future additions
* `Changed` default database cache size of `dm_db_open()` to improve performance

### Programs

* `Added` author option to `dmreport`
* `Added` meta data to PDF output of `dmreport`
* `Fixed` reading of options in `dmdbctl`
* `Changed` last modified date/time to last log date/time in output of `dmfeed`

## [0.9.8] – 2025-08-30

### Library

* `Added` module `dm_image` for image handling
* `Added` module `dm_js` for JavaScript generation
* `Added` module `dm_roff` for GNU roff abstraction
* `Added` module `dm_serial` for derived type serialisation
* `Added` module `dm_transfer` for HTTP-RPC API file transfers
* `Added` GeoJSON Feature Collection procedures to `dm_geojson`
* `Added` image upload to `dm_rpc`
* `Added` HTTP request and response headers to `dm_rpc`
* `Added` HTTP response headers to `dm_fcgi`
* `Added` terminals `gpic`, `postscript`, and `sixeltek` to `dm_plot`
* `Added` image and transfer functions to `dm_db_api`
* `Changed` structure of database abstraction layer, added modules `dm_db_api`, `dm_db_count`, `dm_db_json`, `dm_db_pragma`, and `dm_db_row`
* `Changed` API of modules `dm_arg` and `dm_config` to object-oriented
* `Changed` query parameter API in `dm_cgi`
* `Changed` Git submodules to copies of third-party dependencies
* `Fixed` reading from pipe in `dm_pipe`
* `Fixed` unsigned type conversion in `dm_c`
* `Fixed` return code handling of database select functions

### Programs

* `Added` program `dmcamera` for capturing of camera images
* `Added` program `dmupload` for image upload to HTTP-RPC API
* `Added` scale factor for response values to `dmreport`
* `Added` PDF and PostScript output to `dmreport`
* `Added` message queue reading to `dmmb`
* `Added` image upload to `dmapi`
* `Added` image viewer to `dmweb`
* `Added` environment variables `DM_IMAGE_DB`, `DM_IMAGE_DIR` to `dmapi` and `dmweb`
* `Added` log file output to `dmlogger`
* `Changed` environment variables in `dmapi` and `dmweb` to `DM_BEAT_DB`, `DM_LOG_DB`, `DM_OBSERV_DB`
* `Changed` derived type serialisation in `dmapi` and `dmexport`
* `Changed` database access in `dmexport`
* `Changed` configuration of `dmgrc`

## [0.9.7] – 2025-04-05

### Library

* `Added` module `dm_c` for C interoperability
* `Added` module `dm_camera` for webcam and RTSP access (FFmpeg)
* `Added` module `dm_crypto` of cryptographic hash functions
* `Added` module `dm_db_query` for SQL query building
* `Added` module `dm_dwd` for DWD API format parsing
* `Added` module `dm_dwd_api` for DWD API access (libcurl)
* `Added` module `dm_ftp` for FTP transfer (libcurl)
* `Added` module `dm_freebsd` for system status access on FreeBSD
* `Added` module `dm_geojson` for GeoJSON serialisation
* `Added` module `dm_gm` for GraphicsMagick access
* `Added` module `dm_im` for XMPP connectivity (libstrophe)
* `Added` module `dm_linux` for system status access on Linux
* `Added` module `dm_modbus` for Modbus RTU/TCP connectivity (libmodbus)
* `Added` module `dm_ve` for Victron Energy VE.Direct protocol support
* `Added` library build date and system type parameters (CPP, FPP)
* `Added` Linux and FreeBSD abstractions in module `dm_system`
* `Changed` refactored database abstraction layer
* `Changed` refactored database schema
* `Changed` refactored command-line argument parsing
* `Changed` refactored logical function names
* `Fixed` wrong baud rate in GeoCOM class

### Programs

* `Added` program `dmbot` for remote control of clients through XMPP
* `Added` program `dmdwd` for Deutscher Wetterdienst (DWD) API monitoring
* `Added` program `dmmb` for sensor monitoring through Modbus RTU/TCP
* `Added` program `dmmbctl` for Modbus RTU/TCP command-line control
* `Added` program `dmsystem` for system monitoring
* `Added` program `dmved` for MPPT and battery monitoring (VE.Direct)
* `Added` interactive map view to `dmweb` (Leaflet)

### Documentation


app/dmrecv.f90
! dmrecv.f90
!
! Author:  Philipp Engel
! Licence: ISC
program dmrecv
    !! Receives logs or observations from POSIX message queue and writes them
    !! to standard output or file, either in CSV, JSON Lines, or Namelist
    !! format.
    !!
    !! The output can be piped to a graph tool like _trend(1)_, to show a
    !! real-time plot:
    !!
    !! ```
    !! $ dmrecv --name dmrecv --type observ --format block --response tz0 \
    !!   | awk '{ print $2 | "trend - 60" }'
    !! ```
    !!
    !! Another DMPACK process has to send observations to message queue
    !! `/dmrecv`.  Only responses of name `tz0` will be converted to ASCII
    !! block format and printed to standard output.
    use :: dmpack
    implicit none (type, external)

    character(*), parameter :: APP_NAME  = 'dmrecv'
    integer,      parameter :: APP_MAJOR = 2
    integer,      parameter :: APP_MINOR = 0
    integer,      parameter :: APP_PATCH = 0

    logical, parameter :: APP_MQ_BLOCKING = .true. !! Observation forwarding is blocking.

    type :: app_type
        !! Application settings.
        character(ID_LEN)            :: name        = APP_NAME    !! Name of process and POSIX message queue.
        character(FILE_PATH_LEN)     :: config      = ' '         !! Path to configuration file.
        character(LOGGER_NAME_LEN)   :: logger      = ' '         !! Name of logger (name implies IPC).
        character(NODE_ID_LEN)       :: node_id     = ' '         !! Node id (optional).
        character(FILE_PATH_LEN)     :: output      = ' '         !! Path of output file (stdout if empty or `-`).
        character(FORMAT_NAME_LEN)   :: format_name = ' '         !! Format name.
        character(TYPE_NAME_LEN)     :: type_name   = ' '         !! Type name.
        character(RESPONSE_NAME_LEN) :: response    = ' '         !! Response name for block output of observations.
        integer                      :: format      = FORMAT_NONE !! Data output format.
        integer                      :: type        = TYPE_NONE   !! Data type.
        logical                      :: debug       = .false.     !! Forward debug messages via IPC.
        logical                      :: file        = .false.     !! Output to file.
        logical                      :: forward     = .false.     !! Observation forwarding.
        logical                      :: replace     = .false.     !! Replace output file.
        logical                      :: verbose     = .false.     !! Print debug messages to stderr.
    end type app_type

    class(logger_class), pointer :: logger ! Logger object.

    integer                 :: rc     ! Return code.
    type(app_type)          :: app    ! App settings.
    type(posix_mqueue_type) :: mqueue ! POSIX message queue.

    ! Initialise DMPACK.
    call dm_init()

    ! Get command-line arguments and read options from configuration file.
    rc = read_args(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)

    ! Initialise logger.
    logger => dm_logger_get_default()
    call logger%configure(name    = app%logger,  & ! Name of logger process.
                          node_id = app%node_id, & ! Node id.
                          source  = app%name,    & ! Log source.
                          debug   = app%debug,   & ! Forward debug messages via IPC.
                          ipc     = .true.,      & ! Enable IPC (if logger is set).
                          verbose = app%verbose)   ! Print logs to standard error.
    call logger%info('started ' // APP_NAME)

    init_block: block
        ! Open message queue for reading.
        rc = dm_posix_mqueue_open(mqueue, type=app%type, name=app%name, access=POSIX_MQUEUE_RDONLY)

        if (dm_is_error(rc)) then
            call logger%error('failed to open mqueue /' // app%name, error=rc)
            exit init_block
        end if

        call logger%debug('opened mqueue /' // app%name)

        ! Run the IPC loop.
        call dm_posix_signal_register(signal_callback)
        call run(app, mqueue)
    end block init_block

    call shutdown(rc)
contains
    integer function open_file(path, replace, unit) result(rc)
        character(*), intent(in)  :: path
        logical,      intent(in)  :: replace
        integer,      intent(out) :: unit

        integer :: stat

        rc = E_IO

        if (replace) then
            ! Replace file.
            open (action='write', file=trim(path), iostat=stat, newunit=unit, position='rewind', status='replace')
        else
            ! Append data.
            open (action='write', file=trim(path), iostat=stat, newunit=unit, position='append', status='unknown')
        end if

        if (stat == 0) rc = E_NONE
    end function open_file

    integer function recv_log(app, mqueue) result(rc)
        type(app_type),          intent(inout) :: app    !! App type.
        type(posix_mqueue_type), intent(inout) :: mqueue !! Message queue type.

        character(NML_LOG_LEN) :: log_nml
        integer                :: unit, stat
        type(log_type)         :: log

        unit = stdout
        call logger%debug('waiting for log on mqueue /' // app%name)

        ! Read log from POSIX message queue (blocking).
        rc = dm_posix_mqueue_read(mqueue, log)

        if (dm_is_error(rc)) then
            call logger%error('failed to read log from mqueue /' // app%name, error=rc)
            return
        end if

        if (.not. dm_log_is_valid(log)) then
            call logger%warning('invalid log received', error=E_INVALID)
        else
            call logger%debug('received log ' // trim(log%id))
        end if

        ! Open output file.
        if (app%file) then
            rc = open_file(app%output, app%replace, unit)

            if (dm_is_error(rc)) then
                call logger%error('failed to open file ' // app%output, error=rc)
                return
            end if
        end if

        ! Write serialised log to file/stdout.
        select case (app%format)
            case (FORMAT_CSV)
                ! CSV format.
                rc = dm_csv_write(log, unit=unit)

            case (FORMAT_JSONL)
                ! JSON Lines format.
                rc = dm_json_write(log, unit=unit)

            case (FORMAT_NML)
                ! Namelist format. Write Namelist to a string first, to
                ! avoid newline characters.
                rc = dm_nml_from(log, log_nml)
                write (unit, '(a)', iostat=stat) trim(log_nml)
                if (stat /= 0) rc = E_WRITE
        end select

        ! Close output file.
        if (app%file) close (unit)

        if (dm_is_error(rc)) then
            call logger%error('failed to write log ' // log%id, error=rc)
            return
        end if

        if (app%file) call logger%debug('log ' // trim(log%id) // ' written to ' // app%output)
    end function recv_log

    integer function recv_observ(app, mqueue) result(rc)
        type(app_type),    intent(inout) :: app    !! App type.
        type(posix_mqueue_type), intent(inout) :: mqueue !! Message queue type.

        character(NML_OBSERV_LEN) :: observ_nml
        integer                   :: unit
        integer                   :: idx, stat
        type(dp_type)             :: dp
        type(observ_type)         :: observ

        unit = stdout
        call logger%debug('waiting for observation on mqueue /' // app%name)

        ! Read observation from POSIX message queue (blocking).
        rc = dm_posix_mqueue_read(mqueue, observ)

        if (dm_is_error(rc)) then
            call logger%error('failed to read observation from mqueue /' // app%name, error=rc)
            return
        end if

        if (.not. dm_observ_is_valid(observ)) then
            call logger%warning('invalid observation received', error=E_INVALID)
        else
            call logger%debug('received observation ' // trim(observ%id))
        end if

        ! Open output file.
        if (app%file) then
            rc = open_file(app%output, app%replace, unit)

            if (dm_is_error(rc)) then
                call logger%error('failed to open file ' // app%output, error=rc)
                return
            end if
        end if

        ! Write serialised observation to file/stdout.
        select case (app%format)
            case (FORMAT_BLOCK)
                ! ASCII block format. Search for response of configured name and convert the
                ! observation's response into a data point type.
                idx = dm_observ_find(observ, app%response)

                if (idx > 0) then
                    dp = dp_type(observ%timestamp, observ%responses(idx)%value)
                    rc = dm_block_write(dp, unit=unit)
                else
                    call logger%debug('no response of name ' // app%response, error=E_NOT_FOUND)
                end if

            case (FORMAT_CSV)
                ! CSV format.
                rc = dm_csv_write(observ, unit=unit)

            case (FORMAT_JSONL)
                ! JSON Lines format.
                rc = dm_json_write(observ, unit=unit)

            case (FORMAT_NML)
                ! Namelist format. Write Namelist to a string first, to avoid newline characters.
                rc = dm_nml_from(observ, observ_nml)
                write (unit, '(a)', iostat=stat) trim(observ_nml)
                if (stat /= 0) rc = E_WRITE
        end select

        ! Close output file.
        if (app%file) close (unit)

        if (dm_is_error(rc)) then
            call logger%error('failed to write observation ' // observ%id, error=rc)
            return
        end if

        if (app%file) call logger%debug('observation ' // trim(observ%id) // ' written to ' // app%output)

        ! Forward observation to next receiver.
        if (app%forward) then
            rc = dm_posix_mqueue_forward(observ, name=app%name, blocking=APP_MQ_BLOCKING)
        end if
    end function recv_observ

    subroutine shutdown(error)
        !! Cleans up and stops program.
        integer, intent(in) :: error !! DMPACK error code.

        integer :: rc, stat

        stat = merge(STOP_FAILURE, STOP_SUCCESS, dm_is_error(error))

        call dm_posix_mqueue_close(mqueue, error=rc)
        if (dm_is_error(rc)) call logger%error('failed to close mqueue /' // app%name, error=rc)

        call dm_posix_mqueue_unlink(mqueue, error=rc)
        if (dm_is_error(rc)) call logger%error('failed to unlink mqueue /' // app%name, error=rc)

        call logger%info('stopped ' // APP_NAME, error=error)
        call dm_stop(stat)
    end subroutine shutdown

    subroutine run(app, mqueue)
        !! Event loop that receives logs or observations.
        type(app_type),          intent(inout) :: app    !! App type.
        type(posix_mqueue_type), intent(inout) :: mqueue !! Message queue type.

        integer :: rc

        ipc_loop: do
            select case (app%type)
                case (TYPE_OBSERV); rc = recv_observ(app, mqueue)
                case (TYPE_LOG);    rc = recv_log(app, mqueue)
            end select

            select case (rc)
                case (E_IO);     exit ipc_loop
                case (E_MQUEUE); call dm_sleep(5)
            end select
        end do ipc_loop
    end subroutine run

    ! **************************************************************************
    ! COMMAND-LINE ARGUMENTS AND CONFIGURATION FILE.
    ! **************************************************************************
    integer function read_args(app) result(rc)
        !! Reads command-line arguments and settings from configuration file.
        type(app_type), intent(out) :: app

        type(arg_parser_class) :: parser

        ! Required and optional command-line arguments.
        call parser%add('name',     short='n', type=ARG_TYPE_ID)                            ! -N, --name <string>
        call parser%add('config',   short='c', type=ARG_TYPE_FILE)                          ! -c, --config <path>
        call parser%add('logger',   short='l', type=ARG_TYPE_ID)                            ! -l, --logger <string>
        call parser%add('node',     short='N', type=ARG_TYPE_ID)                            ! -N, --node <string>
        call parser%add('output',   short='o', type=ARG_TYPE_FILE)                          ! -o, --output <path>
        call parser%add('format',   short='f', type=ARG_TYPE_STRING)                        ! -f, --format <string>
        call parser%add('type',     short='t', type=ARG_TYPE_STRING)                        ! -t, --type <string>
        call parser%add('response', short='R', type=ARG_TYPE_ID, max_len=RESPONSE_NAME_LEN) ! -R, --response <string>
        call parser%add('debug',    short='D', type=ARG_TYPE_LOGICAL)                       ! -D, --debug
        call parser%add('forward',  short='F', type=ARG_TYPE_LOGICAL)                       ! -F, --forward
        call parser%add('replace',  short='r', type=ARG_TYPE_LOGICAL)                       ! -r, --replace
        call parser%add('verbose',  short='V', type=ARG_TYPE_LOGICAL)                       ! -V, --verbose

        ! Read all command-line arguments.
        rc = parser%read(version_callback)
        if (dm_is_error(rc)) return

        call parser%get('name',   app%name)
        call parser%get('config', app%config)

        ! Read configuration from file.
        rc = read_config(app)
        if (dm_is_error(rc)) return

        ! Overwrite settings.
        call parser%get('logger',   app%logger)
        call parser%get('node',     app%node_id)
        call parser%get('output',   app%output)
        call parser%get('format',   app%format_name)
        call parser%get('type',     app%type_name)
        call parser%get('response', app%response)
        call parser%get('debug',    app%debug)
        call parser%get('forward',  app%forward)
        call parser%get('replace',  app%replace)
        call parser%get('verbose',  app%verbose)

        app%file   = (dm_string_has(app%output) .and. app%output /= '-')
        app%format = dm_format_from_name(app%format_name)
        app%type   = dm_type_from_name(app%type_name)

        rc = validate(app)
    end function read_args

    integer function read_config(app) result(rc)
        !! Reads configuration from (Lua) file if path is not emty.
        type(app_type), intent(inout) :: app !! App type.

        type(config_class) :: config

        rc = E_NONE
        if (.not. dm_string_has(app%config)) return

        rc = config%open(app%config, app%name)

        if (dm_is_ok(rc)) then
            call config%get('logger',   app%logger)
            call config%get('node',     app%node_id)
            call config%get('output',   app%output)
            call config%get('format',   app%format_name)
            call config%get('response', app%response)
            call config%get('type',     app%type_name)
            call config%get('debug',    app%debug)
            call config%get('forward',  app%forward)
            call config%get('verbose',  app%verbose)
        end if

        call config%close()
    end function read_config

    integer function validate(app) result(rc)
        !! Validates options and prints error messages.
        type(app_type), intent(inout) :: app !! App type.

        rc = E_INVALID

        if (.not. dm_id_is_valid(app%name)) then
            call dm_error_out(rc, 'invalid name')
            return
        end if

        if (dm_string_has(app%node_id) .and. .not. dm_id_is_valid(app%node_id)) then
            call dm_error_out(rc, 'invalid node id')
            return
        end if

        if (dm_string_has(app%logger) .and. .not. dm_id_is_valid(app%logger)) then
            call dm_error_out(rc, 'invalid logger name')
            return
        end if

        if (app%forward .and. app%type /= TYPE_OBSERV) then
            call dm_error_out(rc, '--forward requires type observation')
            return
        end if

        if (app%replace .and. .not. app%file) then
            call dm_error_out(rc, '--replace requires output file')
            return
        end if

        select case (app%format)
            case (FORMAT_BLOCK, FORMAT_CSV, FORMAT_JSONL, FORMAT_NML)
                continue
            case default
                call dm_error_out(rc, 'invalid format')
                return
        end select

        if (app%type /= TYPE_OBSERV .and. app%type /= TYPE_LOG) then
            call dm_error_out(rc, 'invalid type')
            return
        end if

        if (app%format == FORMAT_BLOCK) then
            if (app%type /= TYPE_OBSERV) then
                call dm_error_out(rc, 'block format requires type observation')
                return
            end if

            if (.not. dm_id_is_valid(app%response)) then
                call dm_error_out(rc, 'invalid or missing response name')
                return
            end if
        end if

        rc = E_NONE
    end function validate

    ! **************************************************************************
    ! CALLBACKS.
    ! **************************************************************************
    subroutine signal_callback(signum) bind(c)
        !! C-interoperable signal handler that closes database, removes message
        !! queue, and stops program.
        integer(c_int), intent(in), value :: signum !! Signal number.

        call logger%debug('exit on on signal ' // dm_posix_signal_name(signum))
        call shutdown(E_NONE)
    end subroutine signal_callback

    subroutine version_callback()
        call dm_version_out(APP_NAME, APP_MAJOR, APP_MINOR, APP_PATCH)
        print '(a)', dm_lua_version(.true.)
    end subroutine version_callback
end program dmrecv
